### PR TITLE
Add optional node selector field to NMstate CRD

### PR DESCRIPTION
**Pull request by yboaron**
In current implementation, the handler pods deployed on all the masters and workers nodes.
With node selector, the user can specify on which nodes the handler should run on. in order to do that, the user should first set a label on the relevant nodes and then set the same value in NMstate node selector field.

Signed-off-by: Yossi Boaron yboaron@redhat.com

**Is this a BUG FIX or a FEATURE ?**:

> Uncomment only one, leave it on its own line:
>
> /kind bug
> /kind enhancement

**What this PR does / why we need it**:

**Special notes for your reviewer**:

**Release note**:

```release-note
Add NodeSelector optional field to operator spec, to control on which nodes the handler should run on.
```

### CODE REVIEW

**Comment by qinqon** [comment]
/hold
Are we sure we want this, since we can do same thing at policy level @phoracek ?

**Comment by yboaron** [comment]
Well, yep, with policy we have the flexibility to define on which nodes the config should be applied. 
But for that to happen, we need to run the handler (consume resources on the node, just to check and ignore every policy request)  even on nodes that we don't want to run NMstate. 

I think that these two features are Complementary.

IIUC, the motivation for this feature is avoiding running NMState in nodes that don't satisfy the requirements (NM version and maybe other things), another way to tackle that could be by adding Initcontainer to handler DS (I need to check if it is possible to add initcontainer to DS), and this initcontainer should verify if node satisfies NMState requirements.

**Comment by phoracek** [comment]
We cannot do it via Policies. This is about not running knmstate on some nodes (like masters and incompatible workers) at all. We don't want users to have any chance to run configuration on nodes we know will not work. And we don't want to take those extra resources.

File: pkg/apis/nmstate/v1alpha1/nmstate_types.go [hunk 1]
```diff
@@ -11,6 +11,17 @@ import (
...
+}
+
+// NMstateSpec defines the desired state of NMstate
+type NMstateSpec struct {
+	// ExtraNodeSelector is an optional selector that will be added to handler DaemonSet manifest
+	// for both workers and masters (https://github.com/nmstate/kubernetes-nmstate/blob/master/deploy/handler/operator.yaml).
+	// If ExtraNodeSelector is specified, the handler will run only on nodes that have each of the indicated key-value pairs
+	// as labelsbe applied to the node.
+	// +optional
+	ExtraNodeSelector map[string]string `json:"extranodeselector"`
```
Comment (phoracek) [inline pkg/apis/nmstate/v1alpha1/nmstate_types.go:24]: I think that the extra node is extra. Maybe just NodeSelector? And please make the json to nodeSelector.

Comment (qinqon) [reply pkg/apis/nmstate/v1alpha1/nmstate_types.go:24]: +1

File: pkg/apis/nmstate/v1alpha1/nmstate_types.go [hunk 2]
```diff
@@ -11,6 +11,17 @@ import (
 type NMstate struct {
 	metav1.TypeMeta   `json:",inline"`
 	metav1.ObjectMeta `json:"metadata,omitempty"`
+	Spec              NMstateSpec `json:"spec,omitempty"`
+}
+
+// NMstateSpec defines the desired state of NMstate
+type NMstateSpec struct {
+	// ExtraNodeSelector is an optional selector that will be added to handler DaemonSet manifest
```
Comment (qinqon) [inline pkg/apis/nmstate/v1alpha1/nmstate_types.go:19]: Still `Extra` in the comment

File: deploy/handler/operator.yaml [hunk 3]
```diff
@@ -23,6 +23,9 @@ spec:
       nodeSelector:
         beta.kubernetes.io/arch: amd64
         node-role.kubernetes.io/master: ""
+      {{range $key, $value := .HandlerNodeSelector}}
+        {{$value | printf "%s: %s" $key }}
```
Comment (qinqon) [inline deploy/handler/operator.yaml:27]: What about `{{ $key -}}: {{- $value }}` although dealing with new lines is a pain...

File: pkg/controller/nmstate/nmstate_controller_test.go [hunk 4]
```diff
@@ -69,4 +69,32 @@ var _ = Describe("NMstate controller reconcile", func() {
...
+
+			reconciler.client = cl
+
+			request.Name = existingNMstateName
+		})
+		It("should return empty Result and no error", func() {
+			result, err := reconciler.Reconcile(request)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(result).To(Equal(reconcile.Result{}))
+		})
```
Comment (qinqon) [inline pkg/controller/nmstate/nmstate_controller_test.go:97]: We have to retrieve DaemonSets from handler and check node selector is set there.

File: pkg/controller/nmstate/nmstate_controller_test.go [hunk 5]
```diff
@@ -69,4 +69,32 @@ var _ = Describe("NMstate controller reconcile", func() {
...
 		})
 	})
+	Context("node selector is set ", func() {
+		var (
+			request reconcile.Request
+		)
+		BeforeEach(func() {
+			s := scheme.Scheme
+			s.AddKnownTypes(nmstatev1alpha1.SchemeGroupVersion,
+				&nmstatev1alpha1.NMstate{},
```
Comment (qinqon) [inline pkg/controller/nmstate/nmstate_controller_test.go:79]: maybe we have to add DaemonSet here so we can tretrieve it later to heck nodeSelctor ?

File: pkg/controller/nmstate/nmstate_controller_test.go [hunk 6]
```diff
@@ -69,4 +129,36 @@ var _ = Describe("NMState controller reconcile", func() {
...
+			cl = fake.NewFakeClientWithScheme(s, objs...)
+			reconciler.client = cl
+			request.Name = existingNMStateName
+		})
+		It("NodeSelector labels should be defined for Daemonsets", func() {
+			result, err := reconciler.Reconcile(request)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(result).To(Equal(reconcile.Result{}))
+			// verify NodeSelector value for both worker and master DS
+			ds := &appsv1.DaemonSet{}
```
Comment (qinqon) [inline pkg/controller/nmstate/nmstate_controller_test.go:154]: You can do cl.List and filter by namespace I think

File: pkg/controller/nmstate/nmstate_controller_test.go [hunk 7]
```diff
@@ -16,32 +22,86 @@ import (
...
+	cpCmd = exec.Command("cp", "-f", "../../../deploy/handler/service_account.yaml", "./testdata/kubernetes-nmstate/rbac/")
+	cpCmd.Run()
+	cpCmd = exec.Command("cp", "-f", "../../../deploy/handler/role.yaml", "./testdata/kubernetes-nmstate/rbac/")
+	cpCmd.Run()
+	cpCmd = exec.Command("cp", "-f", "../../../deploy/handler/role_binding.yaml", "./testdata/kubernetes-nmstate/rbac/")
+	cpCmd.Run()
+}
+
+func deleteManifest() {
+	cpCmd := exec.Command("rm", "-f", "./testdata/kubernetes-nmstate/crds/nmstate.io_nodenetworkconfigurationenactments_crd.yaml")
```
Comment (qinqon) [inline pkg/controller/nmstate/nmstate_controller_test.go:45]: You can remove the directory with the golang os module

```golang
os.RemoveAll("/testdata/")
```

File: pkg/controller/nmstate/nmstate_controller_test.go [hunk 8]
```diff
@@ -16,32 +22,86 @@ import (
 	"github.com/nmstate/kubernetes-nmstate/pkg/names"
 )
 
+func copyManifest() {
```
Comment (qinqon) [inline pkg/controller/nmstate/nmstate_controller_test.go:25]: We have to mimic
```
COPY deploy/crds/nmstate.io_nodenetwork*.yaml /bindata/kubernetes-nmstate/crds/
COPY deploy/handler/namespace.yaml /bindata/kubernetes-nmstate/namespace/
COPY deploy/handler/operator.yaml /bindata/kubernetes-nmstate/handler/handler.yaml
COPY deploy/handler/service_account.yaml /bindata/kubernetes-nmstate/rbac/
COPY deploy/handler/role.yaml /bindata/kubernetes-nmstate/rbac/
COPY deploy/handler/role_binding.yaml /bindata/kubernetes-nmstate/rbac/
```
So both CRDs can be use a wildcard

Also can we do a `copyFile` wrapper function to reduce code here ? so we do 
`copyFile("..", "...") 

Also check errors here with `Expect(err).ToNot(HaveOcurred())`

File: pkg/controller/nmstate/nmstate_controller_test.go [hunk 9]
```diff
@@ -16,32 +22,86 @@ import (
...
+
+}
+
+var _ = BeforeSuite(func() {
+	copyManifest()
+})
+
+var _ = AfterSuite(func() {
+	deleteManifest()
+})
```
Comment (qinqon) [inline pkg/controller/nmstate/nmstate_controller_test.go:70]: Can we put this at the _suite_ file ?

File: pkg/controller/nmstate/nmstate_controller_test.go (hunk 6 above)
Comment (yboaron) [reply pkg/controller/nmstate/nmstate_controller_test.go:107]: Yes, it's more elegant to List DS here, but I'm getting [1] error when trying to List DS [2]
Any idea what am I missing here?

[1] tems[0] can't assign or convert unstructured.Unstructured into v1.DaemonSet
[2]
                        dsList := appsv1.DaemonSetList{}
                        err := cl.List(context.TODO(), &dsList, &client.ListOptions{ Namespace: "nmstate"})

File: pkg/controller/nmstate/nmstate_controller_test.go (hunk 8 above)
Comment (yboaron) [reply pkg/controller/nmstate/nmstate_controller_test.go]: I think that I addressed most of your comments here

File: pkg/controller/nmstate/nmstate_controller_test.go (hunk 6 above)
Comment (qinqon) [reply pkg/controller/nmstate/nmstate_controller_test.go:107]: Can we test with client.InNamespace [1] ?

[1] https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/client/client_test.go#L1793

File: pkg/controller/nmstate/nmstate_controller_test.go [hunk 10]
```diff
@@ -69,4 +81,37 @@ var _ = Describe("NMState controller reconcile", func() {
 			Expect(result).To(Equal(reconcile.Result{}))
 		})
 	})
+	Context("node selector is set in Operator spec", func() {
```
Comment (qinqon) [inline pkg/controller/nmstate/nmstate_controller_test.go:84]: "when node ..."

File: pkg/controller/nmstate/nmstate_suite_test.go [hunk 11]
```diff
@@ -9,6 +12,68 @@ import (
...
+		return err
+	}
+	if srcinfo, err = os.Stat(src); err != nil {
+		return err
+	}
+	return os.Chmod(dst, srcinfo.Mode())
+}
+
+func copyManifests() error {
+	dirArr := []string{"crds/", "namespace/", "handler/", "rbac/"}
```
Comment (qinqon) [inline pkg/controller/nmstate/nmstate_suite_test.go:43]: This is not needed is implicit at the srcToDest

File: pkg/controller/nmstate/nmstate_suite_test.go [hunk 12]
```diff
@@ -9,6 +12,68 @@ import (
...
+		"../../../deploy/crds/nmstate.io_nodenetworkconfigurationenactments_crd.yaml": "./testdata/kubernetes-nmstate/crds/",
+		"../../../deploy/crds/nmstate.io_nodenetworkconfigurationpolicies_crd.yaml":   "./testdata/kubernetes-nmstate/crds/",
+		"../../../deploy/crds/nmstate.io_nodenetworkstates_crd.yaml":                  "./testdata/kubernetes-nmstate/crds/",
+		"../../../deploy/handler/namespace.yaml":                                      "./testdata/kubernetes-nmstate/namespace/",
+		"../../../deploy/handler/operator.yaml":                                       "./testdata/kubernetes-nmstate/handler/handler.yaml",
+		"../../../deploy/handler/service_account.yaml":                                "./testdata/kubernetes-nmstate/rbac/",
+		"../../../deploy/handler/role.yaml":                                           "./testdata/kubernetes-nmstate/rbac/",
+		"../../../deploy/handler/role_binding.yaml":                                   "./testdata/kubernetes-nmstate/rbac/",
+	}
+	for _, dir := range dirArr {
```
Comment (qinqon) [inline pkg/controller/nmstate/nmstate_suite_test.go:54]: We can move this to `copyFile` extracting path from destination file no problem on calling MkdirAll if we have already created the dir.

Comment (yboaron) [reply pkg/controller/nmstate/nmstate_suite_test.go]: Yep, good idea, I'll change that.

File: pkg/controller/nmstate/nmstate_suite_test.go [hunk 13]
```diff
@@ -9,6 +12,66 @@ import (
...
+		"../../../deploy/crds/nmstate.io_nodenetworkstates_crd.yaml":                  "./testdata/kubernetes-nmstate/crds/",
+		"../../../deploy/handler/namespace.yaml":                                      "./testdata/kubernetes-nmstate/namespace/",
+		"../../../deploy/handler/operator.yaml":                                       "./testdata/kubernetes-nmstate/handler/handler.yaml",
+		"../../../deploy/handler/service_account.yaml":                                "./testdata/kubernetes-nmstate/rbac/",
+		"../../../deploy/handler/role.yaml":                                           "./testdata/kubernetes-nmstate/rbac/",
+		"../../../deploy/handler/role_binding.yaml":                                   "./testdata/kubernetes-nmstate/rbac/",
+	}
+
+	for src, dest := range srcToDest {
+		if err := os.MkdirAll(dest, os.ModePerm); err != nil {
```
Comment (qinqon) [inline pkg/controller/nmstate/nmstate_suite_test.go:55]: I mean doing `os.MkdirAll` at `copyFile`

**Pull request merged**